components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java

```diff
@@ -4237,6 +4239,54 @@ public static void validateRequestTenantDomain(String tenantDomainOfApp) throws
         }
     }
 
+    /**
+     * Validates whether the tenant domain set in context matches with the app's tenant domain in tenant qualified
+     * URL mode.
+     *
+     * @param tenantDomainOfApp Tenant domain of the app.
+     * @param tokenReqDTO       Access token request DTO object that contains request parameters.
+     * @throws InvalidOAuthClientException
+     */
+    public static void validateRequestTenantDomain(String tenantDomainOfApp, OAuth2AccessTokenReqDTO tokenReqDTO)
+            throws InvalidOAuthClientException {
+
+        if (IdentityTenantUtil.isTenantQualifiedUrlsEnabled()) {
+
+            Optional<String> contextTenantDomainFromTokenReqDTO = getContextTenantDomainFromTokenReqDTO(tokenReqDTO);
+            String tenantDomainFromContext;
+            if (contextTenantDomainFromTokenReqDTO.isPresent() &&
+                    StringUtils.isNotBlank(contextTenantDomainFromTokenReqDTO.get())) {
+                tenantDomainFromContext = contextTenantDomainFromTokenReqDTO.get();
+
+                // In tenant qualified URL mode we would always have the tenant domain in the context.
+                if (!StringUtils.equals(tenantDomainFromContext, tenantDomainOfApp)) {
+                    // This means the tenant domain sent in the request and app's tenant domain do not match.
+                    throw new InvalidOAuthClientException("A valid client with the given client_id cannot be found in "
+                            + "tenantDomain: " + tenantDomainFromContext);
+                }
```

Comment:
shouldn't there any validation for 'else' path here?

Reply:
if the tenant domains are equal, then validation is successful. Therefore, the flow should not be interrupted. The if clause itself is the validation.